src/app.ts

```diff
@@ -3,27 +3,33 @@ import { BigNumberish } from 'ethers';
 import { ChainName, HyperlaneContracts, RouterApp } from '@hyperlane-xyz/sdk';
 import { types } from '@hyperlane-xyz/utils';
 
+import { TokenType } from './config';
 import {
   HypERC20Factories,
   HypERC721Factories,
   TokenFactories,
 } from './contracts';
-import { TokenRouter } from './types';
+import {
+  HypERC20,
+  HypERC20Collateral,
+  HypERC721,
+  HypERC721Collateral,
+  HypNativeCollateral,
+  TokenRouter,
+} from './types';
 
-class HyperlaneTokenApp<
+abstract class HyperlaneTokenApp<
   Factories extends TokenFactories,
 > extends RouterApp<Factories> {
-  router(contracts: HyperlaneContracts<TokenFactories>): TokenRouter {
-    return contracts.router;
-  }
+  abstract router(contracts: HyperlaneContracts<Factories>): TokenRouter;
 
   async transfer(
     origin: ChainName,
     destination: ChainName,
     recipient: types.Address,
     amountOrId: BigNumberish,
   ) {
-    const originRouter = this.getContracts(origin).router;
+    const originRouter = this.router(this.getContracts(origin));
```

Comment:
This fn does not support native collateral, since value should be gasPayment + amountOrId